### ISSUE TITLE
Events for `\Cake\Http\Client`

### ISSUE THREAD

**Issue by asgraf** [cakephp/cakephp#17005]
### Description

I think `\Cake\Http\Client` should emit `beforeSend` and `afterSend` events
Example use cases:
- request caching listener (on beforeSend)
- request throttling listener (on beforeSend)
- request logging listener (on afterSend)

This would be very useful especially when combined with Webservices plugin

### CakePHP Version

5.0 with optional backport for 4.x

**Comment by markstory** [comment]
Why use events? Decorators or 'middleware' seems like an easier to maintain system that doesn't have as much complexity as events do.

**Comment by asgraf** [comment]
Let's take http client logger use case as an example.
Such a logger would need to be attachable globally for all Clients (for example hypothetical global DebugKit logger that logs all Client requests for purpose of displaying them later in DebugKit Toolbar tab) or invidualy per Client instance (log requests from specific client to specific log file)

With event solution we already have global event manager in place where plugins can simply globally attach them selves and with specified priority.

With middleware solution there would need to be two http client middleware queues: global shared queue and local instance queue.

My question for middleware solution is how do you control execution order? Will local middleware queue be called before or after global shared one?

**Comment by markstory** [comment]
> My question for middleware solution is how do you control execution order? Will local middleware queue be called before or after global shared one?

I would expect the globally set middleware to be triggered first. Or we could skip having global/local middleware and use the DI container to prepare wrapped clients that provided to the application.

I am reluctant to add more events. They encourage designing application code that is harder to follow and often contains side-effects that happen far away from the calling code.

Events are also not free. Every event we add and trigger adds overhead to the framework.

**Comment by asgraf** [comment]
I don't insist on events. As long as use cases stated earlier would be possible its 👍 from me

**Comment by markstory** [comment]
In the core team chat @ADmad brought up [PSR-17 factories](https://www.php-fig.org/psr/psr-17/) and it made me think that we *could* have a HTTP client factory that would allow application developers to define 'profiles' for HTTP clients. The factory could work similar to our other configuration based factories. If we wanted to avoid complex configuration data we could use a builder style pattern like:

```php
HttpFactory::set('logging', function (HttpFactory $factory) {
   return new LoggingDecorator(new Client());
});
```

Applications could then use the DI container to get a reference to the application's HTTP factory and request the client they want with `$factory->build('logging');`.  Factories could have a 'fallback' builder that enables the factory to 'just work'. This mode could be disabled as well.

I think this approach would allow applications to build the functionality you're looking for in a declarative fashion. It also separates how the client profiles are configured from how they are used.

What do you think of an approach like this?

**Comment by asgraf** [comment]
Suppose I made a hypothetical application that uses, for example [CVO-Technologies/cakephp-github](https://github.com/CVO-Technologies/cakephp-github/
) plugin
This plugin internally uses the `\Cake\Network\Http\Client` class as every other external API client does
As a person that just uses this external plugin, I would like to be able to add cache, throttle and logging logic to it without any need to modify the source code of this plugin.
With code equivalent to:
```php
//in my bootstrap.php
\Cake\Event\EventManager::instance()
    ->on(
        'Client.beforeSend',
        static function (Client $client, RequestInterface $request, array $options) {
            if ($request->getUri()->getHost() === 'api.github.com') {
                //do some cache logic for github.com
                //do some throttle logic for github.com
            }
        }
    )
    ->on(
        'Client.afterSend',
        static function (Client $client, RequestInterface $request, ResponseInterface $response) {
            if ($request->getUri()->getHost() === 'api.github.com') {
                //do some logging logic for github.com
            }
            //do some logging logic for all hosts
        }
    );
```
Like is said earlier: it does not need to be event based solution.
What is important is that the plugin code should not need to be aware of my additional logic.

**Comment by markstory** [comment]
Thanks for providing that example. With the factory base solution we would need plugins to use named clients so that they could be wrapped. Once a client is created via the factory though, how that client is created or defined is decoupled from where the client is consumed.

Events have an advantage here as they can be applied without any changes to the plugin. In your example you had to use the global event manager to attach listeners. This seems like it would limit the reusability of listeners, as they need conditional logic to filter by host.

**Comment by asgraf** [comment]
Here is more clean & reusable example:
```php
\Cake\Event\EventManager::instance()
    ->on(
        'Client.initialize',
        static function (Client $client) {
            if ($client->getConfig('host') === 'api.github.com') {
                //listeners for api.github.com
                $client->addListener('Cache', [
                    'cacheConfig' => 'default'
                ]);a
                $client->addListener('Throttle', [
                    'period' => 60,
                    'limit' => 100,
                    'cacheConfig' => 'default',
                    'scope' => 'host'//'host' or 'url'
                ]);
                $client->addListener('RequestLog', [
                    'logConfig' => 'client_queries'
                ]);
            }
            //global listeners
            $client->addListener('DebugKit.ClientToolbar');//hypothetical global DebugKit request listener
        }
    );
```
This would also work, since in most API implementations the host is already defined at that point:
https://github.com/CVO-Technologies/cakephp-github/blob/master/src/Webservice/Driver/GitHub.php#L24

**Comment by asgraf** [comment]
I would also like you to check out [dorxy/debug_http](https://github.com/dorxy/debug_http) plugin for cake 3.x
Great plugin but it has two big drawbacks:
- it requires replacing all `\Cake\Http\Client` instances with `\DebugHttp\Network\Http\Client`
- it cannot be used as dev only requirement

Enchantments that im proposing in this ticket would allow to get rid of those drawbacks because `\DebugHttp\Network\Http\Client` class could be simply rewritten as a globally loaded listener

**Comment by markstory** [comment]
> Here is more clean & reusable example:

That is an interesting approach. Having a event when a client is created and then attaching 'middleware' directly to clients lets us build re-usable components more easily and is amenable to DI based solutions where services have factory functions that create their clients. For example:

```php
$container->add(GithubService::class, function (...$args) {
   $client = new Client();
   $client->addListener('Cache', ['config' => 'default']);

   return GithubService($client, ...$args);
});
```

It also avoids much of the overhead that having events on every request would have.

**Comment by asgraf** [comment]
So basically this can be simplified to 2 new methods:
```php
interface ClientListenerInterface
{
    public function addListener(string $className, array $config = []);

    //$urlFilter could be optionally extended to accept `function (Client $client): bool` callable
    public static function addGlobalListener(?string $urlFilter = null, string $className, array $config = []);
}
```
So my previous example could be rewritten to:
```php
//in bootstrap.php
Client::addGlobalListener('api.github.com', 'Cache', ['cacheConfig' => 'default']);
Client::addGlobalListener('api.github.com', 'Throttle', [
    'period' => 60,
    'limit' => 100,
    'cacheConfig' => 'default',
    'scope' => 'host'//'host' or 'url'
]);
Client::addGlobalListener('api.github.com', 'RequestLog', ['logConfig' => 'client_queries']);
Client::addGlobalListener(null, 'DebugKit.ClientToolbar']);
```
From the user's point of view, it does not matter how exactly the `addListener` and `addGlobalListener` methods would be implemented internally and we would maintain full backward compatibility with all currently existing code

**Comment by markstory** [comment]
Is applying  for a single domain that common of a use case? It seems like we'd run into situations where a processor is needed for multiple domains. What do you think of having only an instance method:

```php
public function addProcessor(string | RequestProcessor $class, array $config = [])
```

Attaching to instances created in plugins could be done with the `HttpClient.initialize` event that you outlined before.

I don't think we should use 'listener' in the method name here as we wouldn't be using events under the hood and listener generally means events in CakePHP. I've used 'processor' here, but I don't think that is great either and welcome other ideas.

```php
// Add a processor by string classname.
$client->addProcessor('Cache', ['cacheConfig' => 'http']);

// Add a processor instance
$client->addProcessor($cache);
``` 

Based on our discussion here, we want these processors to be able to manipulate both the outbound request and the response(s) before the response is returned.

I feel like we might be best served by by having listeners/decorators implement an interface similar to `Http\Client\AdapterInterface`. We could have a single method that accepts a request and returns a response. With this approach we can have processors implement logic like:

```php
public function send(
  RequestInterface $request, 
  array $options, 
  ClientRequestHandler $next
): array {
   // Change the request or options before it is sent.
   
   $response = $next->send($request, $options);

   // Change responses before they are returned.
  
   return $response;
}
```

The `ClientRequestHandler` could function similar to the server `RequestHandlerInterface` does by wrapping the queue of processors that has the client's adapter (or mock adapter) as the last step.

**Comment by asgraf** [comment]
> Is applying for a single domain that common of a use case? 

In most cases yes

> It seems like we'd run into situations where a processor is needed for multiple domains.

Thats' why i proposed optionally extending $urlFilter to `callable|string|null` in comment one line above in `ClientListenerInterface` i proposed  in previous post
Example:
```php
$urlFilter = function (Client $client): bool
{
    //if returned value is true Listener/Processor will be applied
    return in_array($client->getConfig('host'), ['domain1', 'domain2', 'domain3']);
};
```


Since `Processors` as you call them are basically middlewares, a question immediately arose in my mind.
Would it make sense that existing middlewares such as [UseMuffin/Throttle](https://github.com/UseMuffin/Throttle) could also be applicable to HttpClient requests and responses? Or am I overthinking this and better keep them as separate things?

**Comment by markstory** [comment]
> Would it make sense that existing middlewares such as [UseMuffin/Throttle](https://github.com/UseMuffin/Throttle) could also be applicable to HttpClient requests and responses? Or am I overthinking this and better keep them as separate things?

There certainly are many similarities, we'll run into some interface incompatibilities because the `ServerRequestInterface` adds several methods to the client request, so static analysis would complain and usage would be likely to have runtime errors. I think having conceptual similarity is still useful though as it means developers have to learn fewer patterns and the conventions they know from response handling middleware is transferable to building client code as well.

> Thats' why i proposed optionally extending $urlFilter to callable|string|null in comment one line above in ClientListenerInterface i proposed in previous post

Ah ok. I didn't pick up on that. What would you think of having the filter callback be provided as an option to `addProcessor()` (or whatever the final name ends up as). Something like:

```php
$client->addProcessor('Cache', [
  'cacheConfig' => 'http',
  'when' => function (ClientRequest $request) {
     return $request->getUri()->getHost() == 'github.com';
   }
]);
```

I don't think there the framework should support `string` values. If filtering by domain becomes common we could provide 'macro' functions like `'when' => ProcessorCondition::matchDomain('api.github.com')` to build these kinds of scenarios

**Comment by asgraf** [comment]
I think the most common filter cases would be:
- single domain:
```php
'api.example.com'
```
- list of domains:
```php
['api.example.com', 'api2.example.com', 'api3.example.com']
```
- wildstar:
```php
'*.example.com'
```
- wildstar list:
```php
['*.example.com/directory1/*', '*.example.com/directory2/*']
```

More complicated cases could be implemented by custom callbacks

**Comment by asgraf** [comment]
Will this be implemented in any shape or form?

**Comment by markstory** [comment]
> Will this be implemented in any shape or form?

I don't need this, so it isn't on my list of things to do. Perhaps someone who needs it will build it.

**Comment by LordSimal** [comment]
As mentioned by my Cakefest workshop this year having events for the Cake HTTP client is also usefull to see which kind of requests and responses are being done by SDK's using that client like the Github SDK (or any other PSR-18 compatible library)

Therefore I will try to implement that.

**Comment by LordSimal** [comment]
At least with the first commit of #17366 my desired functionality is present.

About the ClientFactory/advanced extended functionality let me get my opinion in here as well:

I am not really a fan of adding another middleware-esque system for this kind of purpose especially since its already pretty easy to overwrite the Cake HTTP Client class and extend/overwrite methods themselves.
Having decorators for each usecase and getting those client instances via a Factory surely is a nice abstraction but this all feels a bit too much to do in our HTTP client.

In the end I would rather recommend you use something like guzzle which has this kind of [hooks/middlewares](https://docs.guzzlephp.org/en/stable/handlers-and-middleware.html) already present and go with that. 

As far as I can tell we try to keep our HTTP Client implementation as low level as possible to do the 95% usecases very efficiently without bloating it too much. But this is just (as already mentioned) my opinion on that so if the team wants to go forward with a more rich HTTP Client Implementation then we can do that as well.

**Comment by markstory** [comment]
> I am not really a fan of adding another middleware-esque system for this kind of purpose especially since its already pretty easy to overwrite the Cake HTTP Client class and extend/overwrite methods themselves.
Having decorators for each usecase and getting those client instances via a Factory surely is a nice abstraction but this all feels a bit too much to do in our HTTP client.

That's fair, events are much simpler to implement and understand.

**Comment by LordSimal** [comment]
This feature will be available when CakePHP 5.1 is released.
